evo1/seq/overworld.py

```diff
@@ -74,56 +58,68 @@ def _should_manip(self) -> bool:
-        if self.started_manip:
-            # Check if we've picked up chest, manip done
-            if self.manipulated_enc == self.next_enc:
-                self.started_manip = False
-                self.can_manip = False
-                return False
-        else:
-            # Calculate the manipulated encounter
-            rng = EvolandRNG().get_rng()
-            rng.advance_rng(self._CHEST_RNG_ADVANCE)
-            self.manipulated_enc = calc_next_encounter(rng)
-
-            if not self._should_manip():
+        player = get_zelda_memory().player
+        # Check if we can/should manipulate
+        match self.manip_state:
+            case self._MANIP_FSM.CAN_MANIP:
+                # Check if we are outside range of the chest for next enc
+                dist_to_chest = dist(player.pos, self._CHEST_LOCATION)
+                if not dist_to_chest < player.encounter_timer:
+                    return False
+
+                # Calculate the manipulated encounter
+                rng = EvolandRNG().get_rng()
+                rng.advance_rng(self._CHEST_RNG_ADVANCE)
+                self.manipulated_enc = calc_next_encounter(rng, clink_level=0)
+
+                if not self._should_manip():
+                    return False
+                # Initiate the manip
+                logger.info(
+                    f"Picking up chest to forward rng. {self.manipulated_enc} is better than {self.next_enc}"
+                )
+                self.manip_state = self._MANIP_FSM.STARTED_MANIP
+            case self._MANIP_FSM.STARTED_MANIP:
+                # Move towards chest to manip
+                move_to(
+                    player=player.pos,
+                    target=self._CHEST_LOCATION,
+                    precision=self.precision,
+                )
+                # Check if we've picked up chest, manip done
+                if self.manipulated_enc == self.next_enc or is_close(
+                    player.pos, self._CHEST_LOCATION, precision=self.precision
+                ):
+                    logger.info(f"Manip finished. Next encounter is {self.next_enc}")
+                    self.manip_state = self._MANIP_FSM.MANIP_DONE
+                    return False
+            # Check if we've already picked up the chest
+            case self._MANIP_FSM.MANIP_DONE:
                 return False
 
-        if not self.started_manip:
-            logger.info(
-                f"Picking up chest to forward rng. {self.manipulated_enc} is better than {self.next_enc}"
-            )
-            self.started_manip = True
-
-        # Move towards chest to manip
-        move_to(
-            player=player.pos, target=self._CHEST_LOCATION, precision=self.precision
-        )
-
         return True
 
     def render(self, window: WindowLayout) -> None:
         super().render(window=window)
-        if self.can_manip and self.manipulated_enc:
-            window.stats.addstr(Vec2(1, 15), f"  Manip: {self.manipulated_enc.name}")
```

Comment:
Writing on this line would sometimes cause a crash if it overflowed to the next line, outside the stats window.